packages/juce-blueprint/src/lib/MacroProperties/Transform.ts

```diff
@@ -0,0 +1,30 @@
+import { TPropertyAssignment } from "./types";
+import { getMacroCalls } from "./util";
+
+const rotateMultipliers = {
+  deg: Math.PI / 180.0,
+  grad: Math.PI / 200.0,
+  rad: 1,
+  turn: 2 * Math.PI,
+};
+
+const angleUnitMatcher = new RegExp(
+  `(-?[0-9]*\.?[0-9]+)(${Object.keys(rotateMultipliers).join("|")}|)$`
+);
+
+export default function (value: string): TPropertyAssignment[] {
+  const calls = getMacroCalls(value, ["rotate"]);
```

Comment:
I feel a bit like this won't scale very neatly. I can envision this list growing to include all possible 2D transform macros, and the `if` block on line 19 being duplicated many times for each one. I also think this is a particularly interesting first view property macro, because it ends up being quite complicated.

At the end of the day, each one of these functions boils down to an Affine transform matrix. I'm wondering if our macro helpers here should ultimately return a single affine transform matrix:  `transform-matrix: [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1]` like the `matrix3d` transform function (https://developer.mozilla.org/en-US/docs/Web/CSS/transform).

Then each of these helper functions (rotate, transform, skew, etc) just operate over a matrix that eventually gets passed to native:
```
const transformFunctions = value.match(/([a-z]+\([a-z0-9,.\s]+\))/ig);

// Here transform functions would be a list like [ "rotate(90deg, 25deg)", "translateX(25)" ]
// We reduce over that list, starting with an identity matrix, and ultimately yielding some aggregate matrix
let results = transformFunctions.reduce(function(mat, funcCall) {
   let [funcName, ...args] = splitFuncCall(funcCall);

   // Here funcName would be e.g. "rotate" and "args" would be ["90deg", "25deg"]
   // We call into our `transformFunctionMap` for the `rotate` helper to act over our matrix
   return transformFunctionMap[funcName](mat, ...args);
}, makeIdentityTransformMatrix());

return ['transform-matrix', results];
```

Then we have a map for handling discrete 2d transform function macros.

```
transformFunctionMap.rotate = function(mat, args) {
   // Here we're looking to apply our ["90deg", "25deg"] args over the given `mat` matrix
}
```

The reason I think this is important to do in javascript is because if instead our goal is to ultimately boil down all of these transform helpers to a set of native properties that get set on the View class and handled their, then how do we handle this case?

```css
.hi {
  width:100px;
  height:100px;
  background-color: red;
  transform: rotate(45deg) scale(0.5) rotate(45deg);
}
```
This is a valid transform, and it's equivalent to `rotate(90deg) scale(0.5)`. But if we were to map this to [['transform-rotate', 45], ['scale', 0.5], ['transform-rotate', 45]] you can see that we would clobber the rotate property instead of stacking it.

Reply:
Ah totally yes, this is really the right way to do it in my opinion.  I'll get to that.  And your above comments look good too.

I'm guessing we build should build the full 4d matrix in javascript as suggested, accounting for what's expressible in css, even though I'm not sure we'd use the z dimension in View geometry (just throw away the second-last row/column before handing over to `juce::AffineTransform`).

Reply:
Great!

> I'm guessing we build should build the full 4d matrix in javascript as suggested, accounting for what's expressible in css, even though I'm not sure we'd use the z dimension in View geometry (just throw away the second-last row/column before handing over to juce::AffineTransform).

Yup, totally agreed

Reply:
Hey @nick-thompson , I made another pass through this taking the above into account and it feels way better. Let me know what you think when you have a chance to inspect. Sorry about the editor whitespace changes etc, can pick through them if it's a problem.